src/ns/adapters/outbound/lox24_client.py

```
@@ -0,0 +1,102 @@
+        ),
+    )
+    lox24_sms_send_path: str = Field(
+        default="/sms", description="The path for sending SMS messages"
+    )
+    lox24_auth_token_header: str = Field(
+        default="X-LOX24-AUTH-TOKEN",
+        description="The header for the authentication token",
+    )
+    lox24_sender_id: str = Field(
+        default="GHGA", description="The sender ID to use when sending SMS messages"
+    )
+
+
+class Lox24Client(SmsClientPort):
+    """Concrete implementation of an SmsClientPort for the LOX24 SMS gateway."""
+
+    def __init__(self, *, config: Lox24ClientConfig):
+        """Assign config, which should contain all needed info"""
+        self._config = config
+        self._response: Response | None = None
+        self._send_sms_url: str = f"https://{self._config.lox24_host}:{self._config.lox24_port}/{self._config.lox24_sms_send_path.lstrip('/')}"
+        self._json_data: dict[str, str] = {"sender_id": self._config.lox24_sender_id}
+        self._headers: dict[str, str] = {
+            self._config.lox24_auth_token_header: self._config.lox24_token.get_secret_value()
+        }
+
+    def _raise_for_status(self, response: Response) -> None:
+        """Raise an exception if the response indicates an error."""
+        if response:
+            try:
+                response.raise_for_status()
+            except HTTPStatusError as e:
+                match e.response.status_code:
+                    case 400 | 404 | 422:
+                        raise SmsClientPort.RequestError() from e
+                    case 401 | 402 | 403 | 429:
+                        raise SmsClientPort.AccountError() from e
+                    case 500 | 502 | 503 | 504:
+                        raise SmsClientPort.SystemError() from e
+                    case _:
+                        raise SmsClientPort.GeneralSmsException(
+                            error_info=str(e)
+                        ) from e
```

It would be good to have these errors logged in a way that will help debugging. In the current state, the HTTP errors are obfuscated. The 500-504 errors, for example, result in an error that simply says "Contact Lox24 Support", which doesn't give us a lot to go on.

---

Do you have a preference, should we bubble up the httpx exceptions? Otherwise, I think I need to add a message parameter. The messages are provider-specific and therefore I would prefer to add them in the adapter and not in the port.

---

In `send_sms_message()`, you can wrap `raise_for_status` in a `try/except`, and in the `except` clause, you can log the status code and traceback.

---

Thanks, implemented your suggestion. I added the traceback and logged the response payload since it seems to be helpful. E.g.:

```
ERROR    ns.adapters.outbound.lox24_client:lox24_client.py:107 Received a 401 status code when trying to send SMS. Response payload: {'type': 'https://tools.ietf.org/html/rfc2616#section-10', 'title': 'An error occurred', 'detail': "Client ID or API key isn't active or invalid!"}
Traceback (most recent call last):
  File "/workspace/src/ns/adapters/outbound/lox24_client.py", line 75, in _raise_for_status
    response.raise_for_status()
    ~~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/home/vscode/.local/lib/python3.13/site-packages/httpx/_models.py", line 829, in raise_for_status
    raise HTTPStatusError(message, request=request, response=self)
httpx.HTTPStatusError: Client error '401 Unauthorized' for url 'https://api.lox24.eu/sms/dryrun'
For more information check: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/workspace/src/ns/adapters/outbound/lox24_client.py", line 105, in send_sms_message
    self._raise_for_status(response)
    ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^
  File "/workspace/src/ns/adapters/outbound/lox24_client.py", line 81, in _raise_for_status
    raise SmsClientPort.AccountError() from e
ns.ports.outbound.sms_client.SmsClientPort.AccountError: Authentication failed or account related issue occurred.
```